code/core/src/phone/java/com/adobe/marketing/mobile/services/ui/notification/AEPNotificationUtil.kt

```diff
@@ -0,0 +1,150 @@
+import com.adobe.marketing.mobile.services.ui.notification.builders.AutoCarouselNotificationBuilder
+import com.adobe.marketing.mobile.services.ui.notification.builders.BasicNotificationBuilder
+import com.adobe.marketing.mobile.services.ui.notification.builders.LegacyNotificationBuilder
+import com.adobe.marketing.mobile.services.ui.notification.builders.ManualCarouselNotificationBuilder
+import com.adobe.marketing.mobile.services.ui.notification.models.AEPPushTemplate
+import com.adobe.marketing.mobile.services.ui.notification.models.AutoCarouselPushTemplate
+import com.adobe.marketing.mobile.services.ui.notification.models.BasicPushTemplate
+import com.adobe.marketing.mobile.services.ui.notification.models.CarouselPushTemplate
+import com.adobe.marketing.mobile.services.ui.notification.models.ManualCarouselPushTemplate
+
+/**
+ * Public facing object to construct a [NotificationCompat.Builder] object for the specified [PushTemplateType].
+ * The [constructNotificationBuilder] methods will build the appropriate notification based on the provided
+ * [AEPPushTemplate] or [Intent].
+ */
+object AEPNotificationUtil {
+    private const val SELF_TAG = "AepNotificationUtil"
+
+    @Throws(NotificationConstructionFailedException::class)
+    @JvmStatic
+    fun constructNotificationBuilder(
+        messageData: Map<String, String>,
+        trackerActivityClass: Class<out Activity>?,
+        broadcastReceiverClass: Class<out BroadcastReceiver>?
+    ): NotificationCompat.Builder {
+        val context = ServiceProvider.getInstance().appContextService.applicationContext
+            ?: throw NotificationConstructionFailedException("Application context is null, cannot build a notification.")
+        val pushTemplateType =
+            PushTemplateType.fromString(messageData[PushTemplateConstants.PushPayloadKeys.TEMPLATE_TYPE])
+
+        when (pushTemplateType) {
+            PushTemplateType.BASIC -> {
+                val basicPushTemplate = BasicPushTemplate(messageData as MutableMap<String, String>)
+                return BasicNotificationBuilder.construct(
+                    context,
+                    basicPushTemplate,
+                    trackerActivityClass,
+                    broadcastReceiverClass
+                )
+            }
+
+            PushTemplateType.CAROUSEL -> {
+                val carouselPushTemplate =
+                    CarouselPushTemplate(messageData as MutableMap<String, String>)
+                val carouselOperationMode = carouselPushTemplate.carouselOperationMode
+                val carouselType = carouselPushTemplate.carouselLayoutType
+
+                Log.trace(
+                    PushTemplateConstants.LOG_TAG,
+                    SELF_TAG,
+                    "Building a $carouselType carousel style push notification."
+                )
+
+                return if (carouselOperationMode == PushTemplateConstants.DefaultValues.AUTO_CAROUSEL_MODE) {
```

Comment:
nit: Given the object hierarchy, can you update the creation of carouselPushTemplate and the logic to 

when (carouselPushTemplate) {
    is AutoCarouselPushTemplate -> {
        AutoCarouselNotificationBuilder.construct(
            context,
            carouselPushTemplate,
            trackerActivityClass,
            broadcastReceiverClass
        )
    }
    is ManualCarouselPushTemplate -> {
        ManualCarouselNotificationBuilder.construct(
            context,
            carouselPushTemplate,
            trackerActivityClass,
            broadcastReceiverClass
        )
    }
    else -> 
        NotificationConstructionFailedException
}